packages/@uppy/aws-s3-multipart/src/index.js

```diff
@@ -147,82 +125,118 @@ module.exports = class AwsS3Multipart extends Plugin {
-                ...cFile.s3Multipart.parts,
-                part
-              ]
-            })
-          })
-
-          this.uppy.emit('s3-multipart:part-uploaded', cFile, part)
+      const onSuccess = (result) => {
+        const uploadResp = {
+          uploadURL: result.location
         }
-      }, file.s3Multipart))
+
+        this.resetUploaderReferences(file.id)
+
+        this.uppy.emit('upload-success', file, uploadResp)
+
+        if (result.location) {
+          this.uppy.log('Download ' + upload.file.name + ' from ' + result.location)
+        }
+
+        resolve(upload)
+      }
+
+      const onPartComplete = (part) => {
+        // Store completed parts in state.
+        const cFile = this.uppy.getFile(file.id)
+        if (!cFile) {
+          return
+        }
+        this.uppy.setFileState(file.id, {
+          s3Multipart: {
+            ...cFile.s3Multipart,
+            parts: [
+              ...cFile.s3Multipart.parts,
+              part
+            ]
+          }
+        })
+
+        this.uppy.emit('s3-multipart:part-uploaded', cFile, part)
+      }
+
+      const upload = new Uploader(file.data, {
+        // .bind to pass the file object to each handler.
+        createMultipartUpload: this.requests.wrapPromiseFunction(
+          this.opts.createMultipartUpload.bind(this, file)),
+        listParts: this.requests.wrapPromiseFunction(
+          this.opts.listParts.bind(this, file)),
+        prepareUploadPart: this.opts.prepareUploadPart.bind(this, file),
+        completeMultipartUpload: this.requests.wrapPromiseFunction(
+          this.opts.completeMultipartUpload.bind(this, file)),
+        abortMultipartUpload: this.requests.wrapPromiseFunction(
+          this.opts.abortMultipartUpload.bind(this, file)),
+
+        onStart,
+        onProgress,
+        onError,
+        onSuccess,
+        onPartComplete,
+
+        limit: this.opts.limit || 5,
+        ...file.s3Multipart
+      })
 
       this.uploaders[file.id] = upload
-      this.uploaderEvents[file.id] = createEventTracker(this.uppy)
+      this.uploaderEvents[file.id] = new EventTracker(this.uppy)
+
+      const queuedRequest = this.requests.run(() => {
+        if (!file.isPaused) {
+          upload.start()
+        }
+        // Don't do anything here, the caller will take care of cancelling the upload itself
+        // using resetUploaderReferences(). This is because resetUploaderReferences() has to be
+        // called when this request is still in the queue, and has not been started yet, too. At
+        // that point this cancellation function is not going to be called.
+        return () => {}
+      })
 
       this.onFileRemove(file.id, (removed) => {
+        queuedRequest.abort()
         this.resetUploaderReferences(file.id, { abort: true })
         resolve(`upload ${removed.id} was removed`)
       })
 
+      this.onCancelAll(file.id, () => {
+        queuedRequest.abort()
+        this.resetUploaderReferences(file.id)
```

Comment:
Should it not be `this.resetUploaderReferences(file.id, { abort: true })` here too?

Reply:
yes, thanks!